### PR TITLE
Support php8 by removing safe library

### PR DESCRIPTION
**Pull request by ttomdewit**
1. Remove Safe library
2. Implement Safe methods in own utility class

### CODE REVIEW

File: src/Utilities/SafeMethods.php [hunk 1]
```diff
@@ -4,20 +4,30 @@
...
         int $options = self::JSON_ENCODE_OPTIONS,
         int $depth = self::JSON_ENCODE_DEPTH
     ): string {
-        return \Safe\json_encode(
-            $value,
-            $options,
-            $depth
-        );
+        error_clear_last();
+        $result = \json_encode($value, $options, $depth);
```
Comment (MarijnKoesen) [inline src/Utilities/SafeMethods.php]: You can `use` this as well the other `\` methods below

File: src/Utilities/SafeMethods.php [hunk 2]
```diff
@@ -26,15 +36,27 @@ public function json_decode(
         int $depth = self::JSON_DECODE_DEPTH,
         int $options = self::JSON_DECODE_OPTIONS
     ) {
-        return \Safe\json_decode($json, $assoc, $depth, $options);
+        $data = \json_decode($json, $assoc, $depth, $options);
+        if (JSON_ERROR_NONE !== json_last_error()) {
```
Comment (MarijnKoesen) [inline src/Utilities/SafeMethods.php:60]: I'd try to see if you can get this `json_last_error()` logic to be inside of `JsonException` so you can encapsulate all that logic there, now it's spread across two places. Alternatively you could think about doing something like:

```
throw new JsonException(json_last_error_msg(), json_last_error()); 
```

here instead, to still keep it in 1 place

File: src/Utilities/SafeMethods.php [hunk 3]
```diff
@@ -43,16 +65,40 @@ public function file_put_contents(
...
+            $result = \file_get_contents($filename, $use_include_path, $context, $offset, $maxlen);
+        } elseif (0 !== $offset) {
+            $result = \file_get_contents($filename, $use_include_path, $context, $offset);
+        } elseif (null !== $context) {
+            $result = \file_get_contents($filename, $use_include_path, $context);
+        } else {
+            $result = \file_get_contents($filename, $use_include_path);
+        }
+        if (false === $result) {
+            throw FilesystemException::createFromPhpError();
```
Comment (MarijnKoesen) [inline src/Utilities/SafeMethods.php]: The implementation of this exception requires you to do do a `error_clear_last` before, this is an implementation detail the called should not be aware of, it's leaking responsibilities. 

I think the same as with the other ones, i'd just make that exception get 2 strings in the constructor, and do all that 'magic' here.

```php
public function file_get_contents(..): string {
   error_clear_last();

   [..]

   if (false === $result) {
            $error = error_get_last();
            throw FileSystemException($error['message'] ?? 'An error occured', 0, $error['type'] ?? 1);
   }
}
```

that way it's al nicely done in 1 method.

To make it more explicit you could do:

```php
public function file_get_contents(..): string {
   error_clear_last();

   $result = $this->nativeFileGetContents(...);
   if (false === $result) {
            $error = error_get_last();
            throw FileSystemException($error['message'] ?? 'An error occured', 0, $error['type'] ?? 1);
   }

  return $result;
}

private function nativeFileGetContents(..): mixed {
        if (null !== $maxlen) {
            return \file_get_contents($filename, $use_include_path, $context, $offset, $maxlen);
        } elseif (0 !== $offset) {
            return \file_get_contents($filename, $use_include_path, $context, $offset);
        } elseif (null !== $context) {
            return \file_get_contents($filename, $use_include_path, $context);
        } else {
            return \file_get_contents($filename, $use_include_path);
        }
}
```

This way your exceptions will be agnostic, and all this 'magic' is in the SafeMethods implementation, with a clear (single) responsibility.

File: src/Utilities/SafeMethods.php (hunk 1 above)
Comment (ferrastas) [reply src/Utilities/SafeMethods.php]: +1 and we can automate that via `php-cs-fixer` https://mlocati.github.io/php-cs-fixer-configurator/#version:3.0|fixer:global_namespace_import

**Pull request merged**